Non monomial basis

kalmarek opened the pull request:
This is not even a working code, but a place where we can discuss more concretely

What still needs to be implemented:
 - [x] arithmetic on coefficient vectors
 - [x] decomposition in bases
 - [x] how to encode whether MStructure is of "Dirac type"? (traits?)
 - [x] how to interact with Implicit basis integer indexing seems to be hard in general -- maybe only optional? (again hidden behind a trait? `IndexType`?)
 - [ ] … (and many more)

@blegat

blegat commented on src/bases.jl:
```diff
@@ -1,26 +1,54 @@
+
+Base.eltype(::Type{<:AbstractBasis{T}}) where {T} = T
+Base.keytype(::Type{<:AbstractBasis{T,I}}) where {T,I} = I
+Base.keytype(b::AbstractBasis) = SparseArrays.indtype(typeof(b))
```
The two lines are confusing, I would expect something like
```julia
Base.keytype(::Type{<:AbstractBasis{T,I}}) where {T,I} = I
Base.keytype(b::AbstractBasis) = keytype(typeof(b))
SparseArrays.indtype(::Type{B}) where {B<:AbstractBasis} = keytype(B)
```
to separate the fact that we redirect `b` to its type and the fact that we make `indtype` redirect to `keytype`

blegat commented:
What's failing now are the `MTables` change which I'm less confident with so I hand over the reins back to you

kalmarek commented:
@blegat I'm not trying to  get tests running now. (given the number of planned changes it is a bit pointless., I'm trying to get the implicit/explicit Bases running on my examples;

e.g. with
```julia
G = PermGroup(perm"(1,2,3,4,5,6,7,8)", perm"(1,2)")
db = SA.DiracBasis{UInt32}(G) # <: ImplicitBasis{UInt32, eltype(G)} for an iterable G
mstr = SA.LazyMStructure(db)
g, h = rand(G, 2);
```
and
```julia
function Base.show(io::IO, sc::SA.AbstractCoefficients)
    if iszero(sc)
        print(io, zero(valtype(sc)))
    else
        for (i, (k, v)) in enumerate(pairs(sc))
            print(io, v, '·', k)
            i == length(pairs(sc)) && break
            print(io, " + ")
        end
    end
end
```
I can do now:
```julia
julia> mstr[g, h] == SA.DiracDelta(g * h)
true

julia> x = SA.SparseCoefficients([one(G), g], [1, -1])
1·() + -1·(1,6,5,7,2,8,3,4)

julia> y = sc = SA.SparseCoefficients([one(G), inv(g)], [1, -1])
1·() + -1·(1,4,3,8,2,7,5,6)

julia> res = zero(x)
0

julia> SA.mul!(mstr, res, x, y); res
┌ Error: in __canonicalize!: Not implemented yet
└ @ StarAlgebras ~/.julia/dev/StarAlgebras/src/coefficients.jl:71
1·() + -1·(1,4,3,8,2,7,5,6) + -1·(1,6,5,7,2,8,3,4) + 1·()
```
🎉 Looks good to me even without implementing integer access to `DiracBasis`!

What I don't quite like at the moment is that I had to implement a specialized
```julia
Base.getindex(::LazyMStructure{I,<:DiracBasis{T}}, x::T, y::T) where {I,T}
```
It doesn't feel right...

---

Also: The concept of `I` as indextype for `SA.MultiplicativeStructure{I}` maybe should be tossed out.

blegat commented:
It's working with chebyshev polynomials now, see https://github.com/JuliaAlgebra/MultivariateBases.jl/pull/23
The only issues are

- [ ] I kind of worked around the mtables by directly implementing `fmac!` for the `AblebraElement`. I need to think about it more
- [ ] Didn't do `canonicalize!`. It's actually nontrivial to do this inplace. If you call `sortperm` on the `basis_elements` then you still need to reorder inplace which isn't so easy. It's best to call `sort!` instead but here we have two separate vectors so if we sort one they won't stay in sync. What I do in MultivariatePolynomials is to call `sort!` on a vector of terms. Here, We should create some sort of `AbstractVector{<:Pair}` that we give to `sort!` and for each operation used by `sort!`, we update both vectors

kalmarek commented:
I got both implicit (`DiracBasis`) and explicit (`FixedBasis`) to work;

with some preparation
```julia
using Revise
using StarAlgebras
import StarAlgebras as SA
using PermutationGroups
import PermutationGroups.AP as AP

SA.star(x::Number) = x'
SA.star(g::AP.AbstractPermutation) = inv(g)

struct Lex <: Base.Order.Ordering end
function Base.lt(::Lex, p::AP.AbstractPermutation, q::AP.AbstractPermutation)
    for i in Base.OneTo(min(AP.degree(p), AP.degree(q)))
        ip = i^p
        iq = i^q
        if ip < iq
            return true
        elseif ip > iq
            return false
        end
    end
    return AP.degree(p) < AP.degree(q)
end
Base.isless(p::AP.AbstractPermutation, q::AP.AbstractPermutation) =
    Base.lt(Lex(), p, q)

G = PermGroup(perm"(1,2,3,4,5,6,7,8)", perm"(1,2)") # group of order 40320
```
one can run now:
```julia
db = SA.DiracBasis{UInt32}(G)
RG = SA.StarAlgebra(G, db)

g, h = rand(G, 2)

x = SA.SparseCoefficients([one(G), g], [1, -1])
X = SA.AlgebraElement(x, RG)

y = SA.SparseCoefficients([one(G), inv(g)], [1, -1])
Y = SA.AlgebraElement(y, RG)

xy = SA.SparseCoefficients([one(G), g, inv(g)], [2, -1, -1])
XY = SA.AlgebraElement(xy, RG)

@assert X != Y
@assert X' == Y
@assert RG.mstructure[g, h] == SA.DiracDelta(g * h)
@assert X*Y == XY
```

But also this:
```julia
RGf = let G = G, g = g, h = h
    fb = let (g, h) = (g, h)
        # somehow generate enough elements of G to have
        # a non-trivial multiplication table
        for a in (g, h, inv(g), inv(h), one(g))
            for b in (g, h, inv(g), inv(h), one(g))
                SA.__cache(db, a * b)
            end
        end
        # elts must be star-invariant
        elts = union(db.basis, SA.star.(db.basis))
        SA.FixedBasis{UInt32}(elts)
    end

    k = max(fb[g], fb[h], fb[inv(g)], fb[inv(h)])
    mtbl = SA.MTable(fb, size=(k, k))
    StarAlgebra(G, fb, mtbl)
end

let fb = SA.basis(RGf), g = g
    k = fb[g]
    @assert isone(fb[k] * fb[-signed(k)])
end

using SparseArrays
xf = sparsevec([fb[one(G)], fb[g]], [1, -1], length(fb))
Xf = AlgebraElement(xf, RGf)

yf = sparsevec([fb[one(G)], fb[inv(g)]], [1, -1], length(fb))

Yf = AlgebraElement(yf, RGf)

@assert Xf == Yf'
@assert (Xf')' == Xf
@assert Xf * Xf' == Xf' * Xf
```

kalmarek commented:
Ok, more input/experimentation here. The following fits my usecase (where `AlgebraElements` have their parent objects, but maybe we could tweak it to the Polynomial case as well?)

```julia
f = # polynomial, AlgebraElement, etc, possibly written in different basis
g = # polynomial, AlgebraElement, etc, possibly written in different basis

function +(f, g)
    A = parent(f)
    res = coeffs(f) + coeffs(g, basis(A))
    return AlgebraElement(res, A)
end

function *(f, g)
    A = parent(f)
    mstr = mstructure(A)
    res = mul(mstr, coeffs(f), coeffs(g, basis(A)))
    return AlgebraElement(res, A)
end
```
To spell out assumptions:
* objects `f` have their `parent` object which comes with Explicit/Implicit basis
* `coeffs(g, basis(A))::AbstractCoefficients` decomposes `coeffs(g)` into (possibly sparse) `AbstractCoefficients` written in basis of `A`.
* there is some basic arithmetic on `AbstractCoefficients` over the same basis (but without checking this assumption, by design)
* `mul(mstr, cf, cg)` returns `AbstractCoefficients` which will depend very much on what `cf`, `cg` and `mstr` are. The assumption is (again not checked) that `cf` and `cg` are already written in the basis over which `mstr` is defined.

@blegat I've implemented both implicit, fixed, and augmented (consisting of (1-x)) basis and it seems to be __a__ valid design. what do you think?

blegat commented:
What if `mul(mstr, cf, cg)` becomes `mstr(cf, cg)`

blegat commented:
```julia
function *(f, g)
    A = parent(f)
    mop = moperator(A)
    res = mop(coeffs(f), coeffs(g, basis(A)))
    return AlgebraElement(res, A)
end
```

blegat commented on src/bases.jl:
```diff
@@ -1,26 +1,140 @@
+"""
+abstract type ImplicitBasis{T,I} <: AbstractBasis{T,I} end
+
+mutable struct DiracBasis{T,I,S,St} <: ImplicitBasis{T,I}
```
```suggestion
mutable struct Basis{T,I,S,St} <: ImplicitBasis{T,I}
```

blegat commented on src/bases.jl:
```diff
@@ -1,26 +1,140 @@
+    dict::Dict{T,I}
+    object::S # any iterable
+    state::St
+    lock::Threads.SpinLock
```
Add a field which is `*` by default and contains the multiplicative structure

blegat replied:
`mstructure(::T, ::T)` should be defined

kalmarek commented:
`Basis(itr, moperator=LazyMStructure())`

blegat commented:
`Basis(itr) = Basis(itr, *)`

kalmarek commented:
### things considered as AbstractCoefficients:

`MPoly`
`AugmentedDirac` (1-x)
`SparseVector` # fixed length
`SparseCoefficients` # arbitrary length

blegat commented:
Interface for sparse coefficients implemented by `Dirac`, `SparseCoefficients`, `Vector`, `SparseVector` and `MultivariatePolynomials.AbstractPolynomial`:

* `valtype` and `keytype`
* `nonzero_pairs` returns an iterator of `Pair{K,V}` (`_nzpairs` should be renamed)
* `canonical` returns canonicalized form and `operate!!(canonical, x)` ... We can implement it for `SparseVector` as well
* `UnsafeAddMul` as it currently is, no change needed
* `similar(v, T)` returns a copy and convert the value type

kalmarek commented:
@blegat I migrated everything to `nonzero_pairs`;
There are only a few places where `key_type` is used;

Finally I tried to implement canonicalization using `MA`, but it seems more tricky than I thought so;

kalmarek commented on src/mstructures.jl:
```diff
@@ -15,26 +13,84 @@ function Base.showerror(io::IO, ex::ProductNotDefined)
+        throw(ArgumentError("No alias allowed"))
+    end
+    MA.operate!(zero, res)
+    res = MA.operate!(UnsafeAddMul(ms), res, v, w)
```
@blegat Why do we `MA.operate!` here instead of `MA.operate!!`? We don't know here that `res` is mutable, do we?

kalmarek commented on src/bases_dirac.jl:
```diff
@@ -72,7 +72,7 @@ function coeffs!(
         isone(k) && continue
         x = source[k]
-        unsafe_append!(res, SparseCoefficients((target[AugmentedDirac(x)],), (1,)) => v)
+        MA.operate_to!(res, UnsafeAddMul(*), v, SparseCoefficients((target[AugmentedDirac(x)],), (1,)))
```
@blegat Why do we `MA.operate!` here instead of `res = MA.operate!!`? We don't know here that `res` is mutable, do we?

kalmarek commented:
the failure on macos looks... curious ;)

anyway @blegat there are a bunch of tests marked as `broken` by the non-existent `operate!` for sparse vectors etc. could you have a look at these?

I went on and removed `Dirac` (implemented ala how monomials implement AbstractCoefficients interface).

there is one new addition: `zero_coeffs(::AbstractBasis)` and `[is]one`/`[is]zero` are still wip. also equality etc needs to be tuned.

So far I only converted the old tests. the whole source base needs careful combing to eliminate dead codepaths or the old designs. Not to mention that docs need to be rewritten...

blegat commented:
Thanks! I'll take a look next week and do the next 400 m with the token :)

blegat commented:
I agree, I also prefer using `!` instead of `!!` whenever possible so that I get errors instead of silent performant issues when a mutable method is missing

kalmarek commented:
@blegat somehow codecov stopped working here ~2 weeks ago, do you have idea why?

kalmarek commented:
@blegat I'll try to bring the coverage to reasonable levels, and then merge!

kalmarek commented on test/abstract_coeffs.jl:
```diff
@@ -0,0 +1,100 @@
+
+# for preallocation in arithmetic
+function Base.similar(ac::ACoeffs, ::Type{T}) where {T}
+    # it is needed to zero as uninitialized values may not be cleared by
+    # MA.operate!(zero, ac)
+    return ACoeffs(zeros(T, size(ac.vals)))
```
```julia
vals = similar(ac.vals, T)
MA.operate!(zero, vals)
return ACoeffs(vals)
```

kalmarek commented:
🎉
